### PR TITLE
Closes 222

### PR DESCRIPTION
**Pull request by yywing**
Closes https://github.com/rabbitmq/amqp091-go/issues/222

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -640,9 +640,12 @@ func (c *Connection) dispatch0(f frame) {
...
 		// lolwat - channel0 only responds to methods and heartbeats
-		if err := c.closeWith(ErrUnexpectedFrame); err != nil {
-			Logger.Printf("error sending connectionCloseOk with ErrUnexpectedFrame, error: %+v", err)
-		}
+		// closeWith use call don't block reader
+		go func() {
+			if err := c.closeWith(ErrUnexpectedFrame); err != nil {
+				Logger.Printf("error sending connectionCloseOk with ErrUnexpectedFrame, error: %+v", err)
+			}
+		}()
```
Comment (Zerpet) [inline connection.go:648]: This approach has a lot of potential to cause routine leaks. What's the difference between using a go routine and simply synchronously call the function?

See the following test failure what detected a go routine leak: https://github.com/rabbitmq/amqp091-go/actions/runs/6324272238/job/17185854481?pr=223#step:5:333

File: connection.go [hunk 2]
```diff
@@ -689,9 +692,12 @@ func (c *Connection) dispatchClosed(f frame) {
...
 			// unexpected method on closed channel
-			if err := c.closeWith(ErrClosed); err != nil {
-				Logger.Printf("error sending connectionCloseOk with ErrClosed, error: %+v", err)
-			}
+			// closeWith use call don't block reader
+			go func() {
+				if err := c.closeWith(ErrClosed); err != nil {
+					Logger.Printf("error sending connectionCloseOk with ErrClosed, error: %+v", err)
+				}
+			}()
```
Comment (Zerpet) [inline connection.go:700]: Same comment as before, I suspect this will leak go routines.

**Review by Zerpet — COMMENTED** [review]
Thank you for this PR. I agree with most of the changes. See my comments below for outstanding issues.

File: connection.go (hunk 1 above)
Comment (yywing) [reply connection.go:648]: https://github.com/rabbitmq/amqp091-go/issues/222#issuecomment-1736798083

> n some case, reader will call c.closeWith, and block at c.call at <- c.rpc. But only reader write to <-c.rpc - -

`dispatch0` and `dispatchN` is uesed at `reader` function. If we sync call this fucniton, `reader` will block.

Comment (yywing) [reply connection.go:658]: fixed

Comment (Zerpet) [reply connection.go:658]: I'm happy with the changes in f4ea4b2b288a3863365f021f45a429099e305086

**Comment by Zerpet** [comment]
Thank you!

**Pull request merged**